### PR TITLE
Feature/custom fee logic

### PR DESCRIPTION
**Pull request by antonijzelinskij**
#86959176v

### CODE REVIEW

**Comment by stepanLav** [comment]
Task linked: [CU-86959176v Android. Transfers. Add signed extension for Fees in custom asset (Asset Hub)](https://app.clickup.com/t/86959176v)

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/types/HydrationConversionFeePayment.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    override suspend fun availableCustomFeeAssets(): List<Chain.Asset> {
+        val metaAccount = accountRepository.getSelectedMetaAccount()
+        val chain = chainRegistry.getChain(paymentAsset.chainId)
+        val accountId = metaAccount.accountIdIn(chain)
+        val fromAsset = chain.commissionAsset
+
+        val allSwapDirections = hydraDxQuoteSharedComputation.directions(chain, accountId!!, coroutineScope)
+        val commissionAssetDirections = allSwapDirections.adjacencyList[fromAsset.fullId] ?: emptyList()
+        return commissionAssetDirections.map { chainRegistry.asset(it.direction.to) }
```
Comment (valentunn) [inline feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/types/HydrationConversionFeePayment.kt:55]: Are we sure this will handle a case when there is no direct route between HDX and fee token but there is an inderct route for it? E.g. HDX and USDT:
There is no direct HDX -> USDT route, but there is HDX -> StablePool -> USDT one

I cant recall whether adjencyList returned from hydra swap source includes only direct connections or all possible ones. I feel like the latter should be the case but its better to check

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        chain: Chain,
+        accountId: AccountId,
+        scope: CoroutineScope
+    ): AssetConversion<HydraDxSwapEdge> {
+        val key = "HydraDxAssetConversion:${accountId.toHexString()}"
+
+        return computationalCache.useCache(key, scope) {
+            val storageSharedRequestBuilder = storageSharedRequestsBuilderFactory.create(chain.id)
+            val assetConversion = assetConversionFactory.create(chain)
+            assetConversion.runSubscriptions(accountId, storageSharedRequestBuilder)
```
Comment (valentunn) [inline feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt:72]: Have you tested that subscriptions actually work? We never call `storageSharedRequestBuilder.subscribe` so seems like dont actually subscribe atm. We can subscribe based on AggregateScope that CompotationalCache uses. In that case we pass it to the lambda in `useCache` in case its not already passed there

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    suspend fun quote(
+        chain: Chain,
+        accountId: AccountId,
+        fromAsset: Chain.Asset,
+        toAsset: Chain.Asset,
+        scope: CoroutineScope
+    ): AssetExchangeQuote {
+        val key = "HydraDxQuote:${accountId.toHexString()}"
+
+        return computationalCache.useCache(key, scope) {
```
Comment (valentunn) [inline feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt:47]: This  is the tricky one - quotes may change based on onchain state so we should not cache `assetConversion.quote(paths, args)` if we want to be accurate

File: feature-swap-core/src/test/java/io/novafoundation/nova/ExampleUnitTest.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test
+
+import org.junit.Assert.*
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Comment (valentunn) [inline feature-swap-core/src/test/java/io/novafoundation/nova/ExampleUnitTest.kt:12]: Can be deleted

File: runtime/src/main/java/io/novafoundation/nova/runtime/util/FullAssetIdModel.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package io.novafoundation.nova.runtime.util
+
+import android.os.Parcel
+import android.os.Parcelable
+import io.novafoundation.nova.runtime.multiNetwork.chain.model.ChainId
+import io.novafoundation.nova.runtime.multiNetwork.chain.model.FullChainAssetId
+
+class FullAssetIdModel(val chainId: ChainId, val assetId: Int) : Parcelable {
+
+    constructor(parcel: Parcel) : this(
```
Comment (valentunn) [inline runtime/src/main/java/io/novafoundation/nova/runtime/util/FullAssetIdModel.kt:10]: Cant we use Parcelize here?

File: feature-wallet-impl/src/main/java/io/novafoundation/nova/feature_wallet_impl/data/network/blockchain/assets/transfers/BaseAssetTransfers.kt [hunk 6]
```diff
@@ -48,16 +48,20 @@ abstract class BaseAssetTransfers(
...
     }
 
-    override suspend fun calculateFee(transfer: AssetTransfer): Fee {
-        return extrinsicService.estimateFee(transfer.originChain, TransactionOrigin.SelectedWallet) {
-            transfer(transfer)
-        }
+    override suspend fun calculateFee(transfer: AssetTransfer, coroutineScope: CoroutineScope): Fee {
+        return extrinsicServiceFactory
+            .create(coroutineScope)
+            .estimateFee(transfer.originChain, TransactionOrigin.SelectedWallet) {
```
Comment (valentunn) [inline feature-wallet-impl/src/main/java/io/novafoundation/nova/feature_wallet_impl/data/network/blockchain/assets/transfers/BaseAssetTransfers.kt:62]: Can you please point me where do we pass FeePayment.Asset based on the `transfer`? Also I havent seen a modification of FeeLoaderMixin to account for fee selector, is it going to be a separate PR?

**Review by valentunn — COMMENTED** [review]
Overall approach looks good, we just need to sort of comments related to correctness below

File: feature-account-api/src/main/java/io/novafoundation/nova/feature_account_api/data/model/Fee.kt [hunk 7]
```diff
@@ -23,10 +24,15 @@ data class EvmFee(
     override val amount = gasLimit * gasPrice
 }
 
+interface FeeInAsset : Fee {
+    val assetId: FullChainAssetId
```
[the remark this answers is not in the document]
Comment (antonijzelinskij) [reply feature-account-api/src/main/java/io/novafoundation/nova/feature_account_api/data/model/Fee.kt:28]: I used ids only because it's not very easy to pass Fee using fragment payload (SimpleFeeParcelModel) if we have ChainAsset model inside.

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt (hunk 2 above)
Comment (antonijzelinskij) [reply feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt:72]: Yes it doesn't work. I fixed it in [this](https://github.com/novasamatech/nova-wallet-android/pull/1633) pr

File: runtime/src/main/java/io/novafoundation/nova/runtime/util/FullAssetIdModel.kt (hunk 5 above)
Comment (antonijzelinskij) [reply runtime/src/main/java/io/novafoundation/nova/runtime/util/FullAssetIdModel.kt:10]: Of course. We just need to add `kotlin-android-extensions` plugin

File: feature-account-api/src/main/java/io/novafoundation/nova/feature_account_api/data/model/Fee.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (valentunn) [reply feature-account-api/src/main/java/io/novafoundation/nova/feature_account_api/data/model/Fee.kt:28]: Tbh this looks like half-way solution and will cause inconviniences for us when working with fees

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt [hunk 8]
```diff
@@ -35,31 +36,42 @@ class HydraDxQuoteSharedComputation(
...
         chain: Chain,
+        args: AssetExchangeQuoteArgs,
         accountId: AccountId,
-        fromAsset: Chain.Asset,
-        toAsset: Chain.Asset,
         scope: CoroutineScope
-    ): AssetExchangeQuote {
-        val key = "HydraDxQuote:${accountId.toHexString()}"
+    ): List<Path<HydraDxSwapEdge>> {
+        val key = "HydraDxPaths:${accountId.toHexString()}"
```
Comment (valentunn) [inline feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt:45]: The cache key should reflect all the arguments that might affect results
For example if you call the current code with different asset it will still returned old cached paths

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt [hunk 9]
```diff
@@ -35,31 +36,42 @@ class HydraDxQuoteSharedComputation(
...
             val assetConversion = getAssetConversion(chain, accountId, scope)
             val swapDirections = directions(chain, accountId, scope)
-            val paths = assetConversion.getPaths(swapDirections, args)
-            assetConversion.quote(paths, args)
+            assetConversion.getPaths(swapDirections, args)
         }
     }
 
-    private suspend fun getAssetConversion(
+    suspend fun quote(
```
Comment (valentunn) [inline feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt:54]: Why do we have two `quote` implementations and dublicated code of this implementation in ` HydraDxQuoteSharedComputation`

File: feature-account-api/src/main/java/io/novafoundation/nova/feature_account_api/data/model/Fee.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (antonijzelinskij) [reply feature-account-api/src/main/java/io/novafoundation/nova/feature_account_api/data/model/Fee.kt:28]: I thought that it will be helpful to have separated interface but in UI pr I removed it since it's difficult to support.

Regarding using full chain.asset model in fee: We can create ChainAssetParcel (the same way like ChainParcel using serialization) and pass it to fee payloads. What do you think?

File: feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt (hunk 9 above)
Comment (antonijzelinskij) [reply feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt:54]: Right. Thank you. I forgot to remove this one

**Review by valentunn — APPROVED** [review]
